File: cs/l2_portals.cs
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.NetworkInformation;
using System.Xml.Linq;

namespace l2_portals
{

#nullable enable annotations

class Solution {


    // Quick implementation of PriorityQueue as Meta does not seem to recognize it
    class OurPriorityQueue<TElement, TPriority>  // TODO change this!
        where TPriority : unmanaged
    {
        SortedDictionary<(TPriority, UInt64), TElement> queue;
        UInt64 nb;

        public OurPriorityQueue()
        {
            queue = new SortedDictionary<(TPriority, UInt64), TElement>();
            nb = 0;
        }

        public int Count { get { return queue.Count; } }

        public void Enqueue(TElement element, TPriority priority)
        {
            ++nb;
            queue[(priority, nb)] = element;
        }

        public TElement Dequeue()
        {
            var element = queue.First();
            queue.Remove(element.Key);
            return element.Value;
        }
    }

    class Coord
    {
        public int row;
        public int col;

        public Coord(int _row = 0, int _col = 0)
        {
            row = _row;
            col = _col;
        }
    }

    class NodeInfo : Coord
    {
        public char node_type;
        public int distance;
        public NodeInfo? source;

        public NodeInfo(int _row, int _col, char _node_type)
            : base(_row, _col)
        {
            node_type = _node_type;
            distance = int.MaxValue;
            source = null;
        }
    }

    static void add_neighbour(OurPriorityQueue<NodeInfo, int> q,
        Func<NodeInfo, int> h, Func<NodeInfo, NodeInfo, int> d,
        NodeInfo node, NodeInfo neighbour)
    {
        var neighbor_distance = d(node, neighbour);
        if (neighbor_distance >= neighbour.distance)
            return;
        if (Object.ReferenceEquals(neighbour.source, null))
            q.Enqueue(neighbour, h(neighbour));
        neighbour.source = node;
        neighbour.distance = neighbor_distance;
    }

    public int getSecondsRequired(int R, int C, char[,] G) {
        // https://www.metacareers.com/profile/coding_puzzles/?puzzle=544961100246576
        // Constraints:
        //      1 ≤ R,C ≤ 50
        //      Gi,j ∈ {".", "S", "E", "#", "a"..."z"}
        // Complexity: see A* search algorithm (https://en.wikipedia.org/wiki/A*_search_algorithm)

        // set up grid and portal map
        var grid = new List<List<NodeInfo>>(R);
        for (int j = 0; j < R; ++j)
        {
            var row = new List<NodeInfo>(C);
            grid.Add(row);
            for (int i = 0; i < C; ++i)
                row.Add(new NodeInfo(j, i, G[j, i]));
        }

        var start = new Coord(0, 0);
        var ends = new List<Coord>(R * C);
        var portals = new Dictionary<char, List<NodeInfo>>();
        for (int j = 0; j < R; ++j)
        {
            for (int i = 0; i < C; ++i)
            {
                var node_type = G[j, i];
                if (node_type == 'S')
                    start = new Coord( j, i );
                else if (node_type == 'E')
                    ends.Add(new Coord( j, i ));  // Ends could be used for a heuristic
                else if ('a' <= node_type && node_type <= 'z')
                {
                    if (!portals.ContainsKey(node_type))
                        portals[node_type] = new List<NodeInfo>();  // note: no reserve() here
                    portals[node_type].Add(grid[j][i]);
                }
            }
        }
        var start_node = grid[start.row][start.col];
        start_node.distance = 0;

        // set up the grid
        var q = new OurPriorityQueue<NodeInfo, int>();  // contains (heuristic score, coordinates)
        Func<NodeInfo, int> h = (NodeInfo node) => node.distance;
        Func<NodeInfo, NodeInfo, int> d = (NodeInfo n1, NodeInfo n2) => n1.distance + 1;  // distance is always one

        Coord[] neighbours_directions = { new Coord(-1, 0), new Coord(1, 0), new Coord(0, -1), new Coord(0, 1) };
        var x = h(start_node);
        q.Enqueue(start_node, x);
        while (q.Count > 0)
        {
            var node = q.Dequeue();
            if (node.node_type == 'E')
                return node.distance;
            // add portal nodes to node
            if ('a' <= node.node_type && node.node_type <= 'z')
            {
                foreach (var neighbour in portals[node.node_type])
                    if (!Object.ReferenceEquals(neighbour, node))
                        add_neighbour(q, h, d, node, neighbour);
            }
            // add neighbours to node
            foreach (var dcoord in neighbours_directions)
            {
                var row = node.row + dcoord.row;
                var col = node.col + dcoord.col;
                if ((0 <= row && row < R) && (0 <= col && col < C))
                {
                    var neighbour = grid[row][col];
                    if (neighbour.node_type != '#')
                        add_neighbour(q, h, d, node, neighbour);
                }
            }
        }
        return -1;
    }

    class Args
    {
        public char[,]? G;
        public int res;
    }

    public static int tests()
    {
        Console.WriteLine("\nl2_portals");
        var s = new Solution();
        int nb_errors = 0;

        Func<Args, double> _getSecondsRequired = (Args args) => s.getSecondsRequired(args.G!.GetLength(0), args.G!.GetLength(1), args.G!);

        var args_list = new Args[] {
            new Args { G=new char[,] { { '.', 'E', '.' }, { '.', '#', 'E' }, { '.', 'S', '#' } }, res=4 },
            new Args { G=new char[,] { { 'a', '.', 'S', 'a' }, { '#', '#', '#', '#' }, { 'E', 'b', '.', 'b' } }, res=-1 },
            new Args { G=new char[,] { { 'a', 'S', '.', 'b' }, { '#', '#', '#', '#' }, { 'E', 'b', '.', 'a' } }, res=4 },
            new Args { G=new char[,] { { 'x', 'S', '.', '.', 'x', '.', '.', 'E', 'x' } }, res=3 },
        };

        var nb = 1;
        foreach (Args args in args_list)
        {
            var res = _getSecondsRequired(args);
            if (res == args.res)
                Console.WriteLine("  test #{0}: res={1} CORRECT", nb, res);
            else
            {
                Console.WriteLine("  test #{0}: res={1} ERROR <---------------------", nb, res);
                Console.WriteLine("  expected= {0}", args.res);
                nb_errors += 1;
            }
            ++nb;
        }

        return nb_errors;
    }

}

}

File: js/l2_portals.js
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

// This was converted from Python to Javascript using https://www.codeconvert.ai/
// Conversion did not work immediately (G.map((row, j))
// Then
//      You solved 16 / 31 test cases.
//      Wrong Answer on 15 test cases

class NodeInfo {
    constructor(row, col, nodeType, distance = Number.MAX_SAFE_INTEGER, source = null) {
        this.row = row;
        this.col = col;
        this.nodeType = nodeType;
        this.distance = distance; // cost of the cheapest path from start to n currently known
        this.source = source; // previous node with cheapest path
    }

    debugReconstructPath() {
        let node = this;
        const path = [];
        while (node) {
            path.push(node);
            node = node.source;
        }
        return path.reverse().map(n => [n.col, n.row, n.distance]);
    }
}

function addNeighbour(q, h, d, node, neighbour) {
    const neighborDistance = d(node, neighbour);
    if (neighborDistance >= neighbour.distance) {
        return;
    }
    if (neighbour.source === null) {
        q.push({ score: h(neighbour), node: neighbour });
    }
    neighbour.source = node;
    neighbour.distance = neighborDistance;
}

function getSecondsRequired(R, C, G) {
    // https://www.metacareers.com/profile/coding_puzzles/?puzzle=544961100246576
    // Constraints:
    //      1 ≤ R,C ≤ 50
    //      Gi,j ∈ {".", "S", "E", "#", "a"..."z"}
    // Complexity: see A* search algorithm (https://en.wikipedia.org/wiki/A*_search_algorithm)

    // codeconvert does not "map" correctly string
    //const grid = G.map((row, j) => row.map((elt, i) => new NodeInfo(j, i, elt)));
    const grid = G.map((row, j) => Array.from(row).map((elt, i) => new NodeInfo(j, i, elt)));
    const portals = {};

    let start = [0, 0];
    const ends = [];
    for (let j = 0; j < G.length; j++) {
        for (let i = 0; i < G[j].length; i++) {
            const nodeType = G[j][i];
            if (nodeType === 'S') {
                start = [j, i];
            } else if (nodeType === 'E') {
                ends.push([j, i]); // Ends could be used for a heuristic
            } else if (nodeType >= 'a' && nodeType <= 'z') {
                if (!portals[nodeType]) portals[nodeType] = [];
                portals[nodeType].push(grid[j][i]);
            }
        }
    }
    const startNode = grid[start[0]][start[1]];
    startNode.distance = 0;

    // codeconvert uses a list, sorted multiple type as javascript does not have a min or max heap
    const q = []; // contains { score, node }
    const h = n => n.distance; // heuristic is the node distance (very basic)
    const d = (n1, n2) => n1.distance + 1; // distance is always one

    const neighboursDirections = [[-1, 0], [1, 0], [0, -1], [0, 1]];

    q.push({ score: h(startNode), node: startNode });
    while (q.length) {
        // codeconvert did not pick up NodeInfo.__lt__ in the Python source
        //q.sort((a, b) => a.score - b.score);  // codeconvert: You solved 16 / 31 test cases.
        q.sort((a, b) => (a.score - b.score, a.node.distance - b.node.distance));
        const { score, node } = q.shift();
        if (node.nodeType === 'E') {
            return node.distance;
        }
        // add portal nodes to node
        if (node.nodeType >= 'a' && node.nodeType <= 'z') {
            for (const neighbour of portals[node.nodeType]) {
                if (neighbour !== node) {
                    addNeighbour(q, h, d, node, neighbour);
                }
            }
        }
        // add neighbours to node
        const neighboursCoords = neighboursDirections.map(([drow, dcol]) => [node.row + drow, node.col + dcol]);
        const neighbours = neighboursCoords
            .filter(([row, col]) => (0 <= row && row < R) && (0 <= col && col < C))
            .map(([row, col]) => grid[row][col]);
        for (const neighbour of neighbours) {
            if (neighbour.nodeType !== '#') {
                addNeighbour(q, h, d, node, neighbour);
            }
        }
    }
    return -1;
}

function tests() {
    const fn = G => [G.length, G[0].length, G];
    const metaCases = ["meta", [
        [[[".E.", ".#E", ".S#"]], 4],
        [[["a.Sa", "####", "Eb.b"]], -1],
        [[["aS.b", "####", "Eb.a"]], 4],
        [[["xS..x..Ex"]], 3],
    ]];
    return [getSecondsRequired, fn, [metaCases]];
}

module.exports = {tests};
